qadence/backends/adjoint.py

```diff
@@ -0,0 +1,117 @@
+from qadence.blocks.abstract import AbstractBlock
+
+
+class AdjointExpectation(Function):
+    @staticmethod
+    @no_grad()
+    def forward(
+        ctx: Any,
+        circuit: PyQCircuit,
+        observable: PyQCircuit,
+        state: Tensor,
+        param_names: list[str],
+        *param_values: Tensor,
+    ) -> Tensor:
+        for param in param_values:
+            param = param.detach()
+        ctx.circuit = circuit
+        ctx.observable = observable
+        ctx.param_names = param_names
+        values = param_dict(param_names, param_values)
+        ctx.out_state = circuit.run(state, values)
+        ctx.projected_state = observable.run(ctx.out_state, values)
+        ctx.save_for_backward(*param_values)
+        return overlap(ctx.out_state, ctx.projected_state)
+
+    @staticmethod
+    def backward(ctx: Any, grad_out: Tensor) -> tuple:
+        param_values = ctx.saved_tensors
+        values = param_dict(ctx.param_names, param_values)
+
+        def _apply_adjoint(ctx: Any, op: Module) -> list:
+            grads: list = []
+            if isinstance(op, (PyQHamiltonianEvolution)):
+                generator = op.block.generator
+                time_param = values[op.param_names[0]]
+
+                ctx.out_state = apply_operator(ctx.out_state, op.dagger(values), op.qubit_support)
+                if (
+                    isinstance(generator, AbstractBlock)
+                    and generator.is_parametric
+                    and values[op.param_names[1]].requires_grad
+                ):
+                    mu = apply_operator(
+                        ctx.out_state, op.jacobian_generator(values), op.qubit_support
+                    )
+                    grads.append(2 * overlap(ctx.projected_state, mu))
+                elif time_param.requires_grad:
+                    mu = apply_operator(ctx.out_state, op.jacobian_time(values), op.qubit_support)
+                    grads.append(2 * overlap(ctx.projected_state, mu))
+                ctx.projected_state = apply_operator(
+                    ctx.projected_state, op.dagger(values), op.qubit_support
+                )
+            elif isinstance(op, ScalePyQOperation):
+                ctx.out_state = apply_operator(ctx.out_state, op.dagger(values), op.qubit_support)
+                scaled_pyq_op = op.operations[0]
+                if (
+                    isinstance(scaled_pyq_op, PyQParametric)
+                    and values[scaled_pyq_op.param_name].requires_grad
+                ):
+                    mu = apply_operator(
+                        ctx.out_state,
+                        scaled_pyq_op.jacobian(values),
+                        scaled_pyq_op.qubit_support,
+                    )
+                    grads.append(2 * overlap(ctx.projected_state, mu))
+
+                if values[op.param_name].requires_grad:
+                    grads.append(2 * -values[op.param_name])
```

Comment:
@dominikandreasseitz where does the minus sign come from? I would expect a plus sign here but I might need to re-read the paper.

Reply:
scaleblock derivatives are not in the paper, i computed it somehow, should be correct. there is also a test ofc

Reply:
would resolve this since i verified it gives the right result